### ISSUE TITLE
Receive "Deployment bucket is not in the same region as the lambda function" when using us-east-1

### ISSUE THREAD

**Issue by briguy202** [serverless/serverless#2404]
# This is a Bug Report
## Description
- What went wrong?

When attempting to use a pre-existing S3 bucket in us-east-1 via the `deploymentBucket` configuration option, I receive an error response "Deployment bucket is not in the same region as the lambda function".

Debugging through the Serverless code, I observed that it is attempting to match a LocationConstraint returned from the AWS SDK to the region; however, [AWS Documentation](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETlocation.html) states: 

> When the bucket's region is US East (N. Virginia), Amazon S3 returns an empty string for the bucket's region

Due to this, it appears the check in **configureStack.js** to match the bucket's region with the specified region fails and I receive the error message.
- What did you expect should have happened?

I should have been able to use the pre-existing bucket.
- What was the config you used?

```
...
provider:
  name: aws
  runtime: nodejs4.3
  stage: dev
  region: us-east-1
  deploymentBucket: my-bucket-name
...
```
- What stacktrace or error message from your provider did you see?

Deployment bucket is not in the same region as the lambda function
## Additional Data
- **_Serverless Framework Version you're using**_: 1.0.2
- **_Operating System**_: OSX
- **_Stack Trace**_:
- **_Provider Error messages**_:

```
  Serverless Error ---------------------------------------

     Deployment bucket is not in the same region as the lambda
     function.
```

**Comment by eahefnawy** [comment]
Thanks for reporting and investigating @briguy202 😊 .... I've pushed a fix along with [this PR](https://github.com/serverless/serverless/pull/2385) and will be merged very soon. It's a high priority!

Cheers 👼

**Comment by doapp-ryanp** [comment]
Is there a workaround for this one? I was hoping that omitting `provider.region` would get around it but no dice.

**Comment by briguy202** [comment]
@doapp-ryanp, unfortunately I don't think there is other than to patch the issue or wait on #2385 which looks to be very close to completion.  If you opt for patching, which is what we did, it's a one-liner change - see https://github.com/briguy202/serverless/commit/750c55b275ba5a022479154dfd89f26092818682.

**Comment by pmuens** [comment]
Should be resolved with #2385 /cc @flomotlik

**Comment by ssemyonov** [comment]
I've got the same issue happening when using pre-existing deployment bucket in ```eu-west-1``` region in some of AWS accounts in my company. The reason is AWS returns ```LocationConstraint``` value as ```EU``` which is one of the values for ```eu-west-1``` according to [AWS docs](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETlocation.html):

> Valid Values: [ us-west-1 | us-west-2 | ca-central-1 | **EU or eu-west-1** | eu-west-2 | eu-central-1 | ap-south-1 | ap-southeast-1 | ap-southeast-2 | ap-northeast-1 | ap-northeast-2 | sa-east-1 | empty string (for the US East (N. Virginia) region) | us-east-2]

This could be one-liner fix in **configureStack.js** similar to the fix for original issue:
```javascript
if (result.LocationConstraint === 'EU') result.LocationConstraint = 'eu-west-1';
```

Should I raise a separate issue for this case or current one could be reopened?

**Comment by eahefnawy** [comment]
BOOM! 💥 That is one nice catch @ssemyonov ... I feel like we're always on a mission to keep up with AWS inconsistency 😄 ... I'm gonna PR a quick fix today.

**Comment by ssemyonov** [comment]
Hi @eahefnawy are you still looking into EU - eu-west-1 issue? Do you need a hand?

**Comment by ssemyonov** [comment]
PR #3107 created to address EU - eu-west-1 issue.

**Comment by voiceactivity** [comment]
For some reasons this merge was lost. Having same issue with the latest sls -> 1.27.2
Can confirm that at least master lost this bit:

if (result.LocationConstraint === 'EU') result.LocationConstraint = 'eu-west-1';

You can check this out here: https://github.com/ssemyonov/serverless/blob/master/lib/plugins/aws/deploy/lib/configureStack.js [line 33] 

  Serverless Error ---------------------------------------

  Could not locate deployment bucket. Error: Deployment bucket is not in the same region as the lambda function

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.9.1
     Serverless Version:     1.27.2

Could you please fix it!

**Comment by HyperBrain** [comment]
@voiceactivity The check is still there - in `existsDeploymentBucket.js`:

```
        const result = resultParam;
        if (result.LocationConstraint === '') result.LocationConstraint = 'us-east-1';
        if (result.LocationConstraint === 'EU') result.LocationConstraint = 'eu-west-1';
        if (result.LocationConstraint !== this.provider.getRegion()) {
```
https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/lib/existsDeploymentBucket.js#L17

It seems that you have some old code there - the configureStack file has been removed for some time now 😄

**Comment by voiceactivity** [comment]
Ok, but this fact doesn't explain why I am having that error with us-east-1 region.
This is my serverless.yml:

provider:
  name: aws
  runtime: nodejs8.10
  stage: ${opt:stage, self:custom.defaultStage}
  profile: ${self:custom.profiles.${self:provider.stage}}
  region: us-east-1
  memorySize: 256
  timeout: 10
  role: ${file(./env.config.${self:provider.stage}.json):LAMBDA_ROLE}
  deploymentBucket:
    name: xyz-lambdas-${self:provider.stage}    
custom:
...
functions:
  myfunction:
    handler: index.handler
    environment:
      ALEXA_SKILL_VENDOR_ID: ${file(./env.config.${self:provider.stage}.json):ALEXA_SKILL_VENDOR_ID}
    events:
      - alexaSkill: 
        appId: amzn1.ask.skill.xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx
        enabled: true
...

The source of pain is 'deploymentBucket'. When I removed this property then the serverless creates the S3 bucket for me with a random name. That works! But, as soon as I am using the 'deploymentBucket' property it giving up and shows me that error.
Could that be that if the 'deploymentBucket' name provided then some other validation used?

**Comment by voiceactivity** [comment]
Thanks for that. I am still getting the error. I will investigate  further.

**Comment by priyaprabhakar777** [comment]
May I know why it is not allowed to use the S3 bucket from different region? Is there any reason for that?

**Comment by max-lobur** [comment]
Having the same issue when using profiles. Profile has `us-west-2` region set, and bucket is in `us-west-2`, however I get `Deployment bucket is not in the same region as the lambda function`. AWS Env vars are not set, so I'm not sure where it takes the other region.

Setting `us-west-2` in provider explicitly works, but that breaks the profiles abstraction :(

**Comment by rodrigogs** [comment]
Having it trying to run `sls dev` for `sa-east-1`. No clue on how to solve this.

**Comment by seba-i** [comment]
For the sake of trying it, commenting out the location constraint worked for my simple resource (SES templates) deploy file.

node_modules/serverless/lib/plugins/aws/deploy/lib/existsDeploymentBucket.js
```
  ...
  const result = resultParam;
  if (result.LocationConstraint === '') result.LocationConstraint = 'us-east-1';
  if (result.LocationConstraint === 'EU') result.LocationConstraint = 'eu-west-1';
  // if (result.LocationConstraint !== this.provider.getRegion()) {
  //   throw new ServerlessError(
  //     'Deployment bucket is not in the same region as the lambda function'
  //   );
  // }
  return BbPromise.resolve();
  ...
```
Buckets are global, why is this constraint necessary? Unless you've lucked out and created everything in a region where every service you'll ever need is available, its likely to cause a headache some day...
